cMake/FreeCAD_Helpers/CompilerChecksAndSetups.cmake

```diff
@@ -67,6 +75,44 @@ macro(CompilerChecksAndSetups)
         endif()
         include_directories(${CMAKE_CURRENT_BINARY_DIR})
 
+        set(_FREECAD_STL_ASSERT_GENEX_DEBUG $<AND:$<CONFIG:Debug>,$<COMPILE_LANGUAGE:CXX>>)
+        set(_FREECAD_STL_ASSERT_GENEX_ALL   $<COMPILE_LANGUAGE:CXX>)
+
+        if(FREECAD_STL_ASSERTIONS STREQUAL "ON")
+            set(_FREECAD_STL_ASSERT_GENEX ${_FREECAD_STL_ASSERT_GENEX_ALL})
+            set(_FREECAD_STL_ASSERT_SCOPE "all configs")
+        elseif(FREECAD_STL_ASSERTIONS STREQUAL "AUTO")
+            set(_FREECAD_STL_ASSERT_GENEX ${_FREECAD_STL_ASSERT_GENEX_DEBUG})
+            set(_FREECAD_STL_ASSERT_SCOPE "Debug only")
+        else()
+            set(_FREECAD_STL_ASSERT_GENEX "")  # OFF
+            set(_FREECAD_STL_ASSERT_SCOPE "disabled")
+        endif()
+
+        if(_FREECAD_STL_ASSERT_GENEX)
+            # Detect the STL implementation in use
+            include(CheckCXXSourceCompiles)
+            set(_FREECAD_PROBE_LIBSTDCXX "#include <vector>\nint main() {\n#ifdef __GLIBCXX__\n  return 0;\n#else\n# error not libstdc++\n#endif\n}")
+            check_cxx_source_compiles("${_FREECAD_PROBE_LIBSTDCXX}" _FREECAD_USE_LIBSTDCXX)
+
+            set(_FREECAD_PROBE_LIBCXX "#include <vector>\nint main() {\n#ifdef _LIBCPP_VERSION\n  return 0;\n#else\n# error not libc++\n#endif\n}")
+            check_cxx_source_compiles("${_FREECAD_PROBE_LIBCXX}" _FREECAD_USE_LIBCXX)
+
+            if(_FREECAD_USE_LIBSTDCXX)
+                add_compile_definitions($<${_FREECAD_STL_ASSERT_GENEX}:_GLIBCXX_ASSERTIONS>)
+                message(STATUS "STL assertions: libstdc++ _GLIBCXX_ASSERTIONS (${_FREECAD_STL_ASSERT_SCOPE})")
+            elseif(_FREECAD_USE_LIBCXX OR BUILD_USE_LIBCXX)
+                add_compile_definitions(
+                    $<${_FREECAD_STL_ASSERT_GENEX}:_LIBCPP_ENABLE_ASSERTIONS=1>
```

Comment:
Note that `_LIBCPP_ENABLE_ASSERTIONS` was deprecated in LLVM https://github.com/llvm/llvm-project/pull/113592